### PR TITLE
fix(combobox): modal focus issue

### LINKED ISSUE #21153: [Bug]: Tabs do not activate in Modal sometimes
### Package

@carbon/react

### Browser

Firefox

### Package version

v1.95.0

### React version

v19.1.1

### Description

When using Carbon’s Tabs inside a Modal, the tab doesn’t activate on click near the edges of the view. Clicking the same tab again now selects the tab.

See example with Vertical Tabs (issue also reproducible with horizontal tabs)

**Current behaviour:**
Currently, the tab auto-scrolls to the center on click, but isn't active.

**Expected behaviour:**
The tab auto-scrolls to the center and is active.

### Reproduction/example

https://stackblitz.com/edit/github-t39n99kg?file=src%2FApp.jsx

### Steps to reproduce

1. Open Preview in a new tab.
2. Open modal.
3. Cilck on an inactive tab that is closer to the top or bottom border.
4. Tab scrolls to center, has `:focus` pseudo-class, but is not active.
5. Press the tab again to activate.

### Suggested Severity

Severity 3 = User can complete task, and/or has a workaround within the user experience of a given component.

### Application/PAL

_No response_

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [x] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

### LINKED ISSUE #21638: [Bug]: Combobox within modal focus issue
### Package

@carbon/react

### Browser

Chrome

### Package version

1.97.0

### React version

18.3.1

### Description

When there is a combobox within a modal and within tabs, when user try to clear the default value in the combobox, the modal scroll instead of clearing the modal. User have to click the X button twice to clear the combobox

### Reproduction/example

https://stackblitz.com/edit/eumm8zsk-bxbindrs?file=src%2FApp.jsx

### Steps to reproduce

open the modal and try to clear the combobx

### Suggested Severity

Severity 3 = User can complete task, and/or has a workaround within the user experience of a given component.

### Application/PAL

FTM

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [x] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

### PR DESCRIPTION
**Pull request by gDINESH13**
Closes #21638
Closes #21153

When clicking interactive elements (ComboBox clear button, tabs) inside a Modal, the Modal's auto-scroll behavior interrupts the click event, requiring users to click twice - once to scroll the element into view, and again to actually trigger the action.

Cause:

The Modal's `wrapFocus` handler calls `scrollIntoView` when an element receives focus, which happens during the click event. This scroll interrupts the click before it completes.

### Changelog

**New**

- None

**Changed**

1. *Skip auto-scroll for buttons*: Buttons receive focus from explicit user clicks and should never auto-scroll
2. *Skip auto-scroll for visible elements*: If an element is already fully visible, scrolling is unnecessary and disruptive

**Removed**

- None

#### Testing / Reviewing

I haven't added test case, it was tricky to emulate test cases, with scrolling scenarios. However I have created stories with the slack bits examples shared in the issues and have tested the functionality after the fix. 
I have added screen recoding of testing in the comments, please take a look 😄

## PR Checklist

As the author of this PR, before marking ready for review, confirm you:

- [x] Reviewed every line of the diff
~- [] Updated documentation and storybook examples~
~- [] Wrote passing tests that cover this change~
~- [] Addressed any impact on accessibility (a11y)~
- [x] Tested for cross-browser consistency
- [x] Validated that this code is ready for review and status checks should pass

More details can be found in the [pull request guide](https://github.com/carbon-design-system/carbon/blob/main/docs/guides/reviewing-pull-requests.md)

### CODE REVIEW

**Review by adamalston — COMMENTED** [review]
I reverted the fix, and the test still passes, which suggests that it is not covering the changes.

```sh
yarn test packages/react/src/components/ComboBox/ComboBox-test.js
git checkout main packages/react/src/components/ListBox/next/ListBoxSelection.tsx packages/react/src/internal/wrapFocus.ts
yarn test packages/react/src/components/ComboBox/ComboBox-test.js
```

**Comment by gDINESH13** [comment]
Hi @adamalston thankyou for the review. Is it fine if I add a new story in combobox-test to see if it fixed the issue? I can revert the example later, after review. Or if I create a new story in combobox ( just for testing & reviewing ), it shall be moved together with this PR?

**Comment by adamalston** [comment]
Hi, I think there may be a misunderstanding. My point was that if I revert the fix, the existing test still passes, which means the test isn't actually covering the issue this pull request intends to fix.

Ideally we should have a test that fails without the fix and passes with it. That way we know the behavior is actually being validated.

Adding a Storybook example could help demonstrate the issue, but it wouldn't replace adding a test that reproduces the issue.

**Comment by gDINESH13** [comment]
Hi @adamalston sorry it's been a while since ur comment. Initially I have raised this PR only for #21638 but later I saw @maradwan26 pointing the same problem in other issue #21153, did some research and I think the problem is in Modal component, I thought its an issue on combobox initially.

I have updated the PR description with new changes. 

Adding the screen recordings of SlackBits mentioned in issues,
 
Tabs 

https://github.com/user-attachments/assets/3bf48ad6-acd6-4cd3-a9cc-79cfa0da1148



Combo box

https://github.com/user-attachments/assets/5d777e37-345e-42fd-8742-a7e28ba5c171

I haven't added story book examples in the commit, If guys feel it's better to have them in repo, I don't mind adding later. Please let me know what you guys think. Thanks.

File: packages/react/src/components/Modal/Modal.tsx [hunk 1]
```diff
@@ -463,6 +463,28 @@ const ModalDialog = React.forwardRef(function ModalDialog(
...
 
+    // Check if the focused element is a button - buttons should never auto-scroll
+    // as clicking them is an explicit user action that shouldn't be interrupted
+    const isButton = currentActiveNode.matches(
+      'button, [role="button"], input[type="button"], input[type="submit"]'
+    );
+
+    if (isButton) {
+      return; // Never auto-scroll buttons
+    }
```
Comment (adamalston) [inline packages/react/src/components/Modal/Modal.tsx]: Should this early return be scoped to focus changes caused by clicking or tapping, or otherwise keep keyboard focused buttons scrollable into view? With the blanket button check here, if you tab to a `button` or `[role="button"]` inside a long modal body, focus can stay below the visible area under the fade.

File: packages/react/src/components/Modal/Modal.tsx [hunk 2]
```diff
@@ -463,6 +463,28 @@ const ModalDialog = React.forwardRef(function ModalDialog(
...
+    const elementRect = currentActiveNode.getBoundingClientRect();
+
+    // Check if element is fully visible within the modal viewport
+    const isFullyVisible =
+      elementRect.top >= modalRect.top &&
+      elementRect.bottom <= modalRect.bottom;
+
+    if (isFullyVisible) {
+      return; // Don't scroll if already fully visible
+    }
```
Comment (adamalston) [inline packages/react/src/components/Modal/Modal.tsx:476]: How is the faded bottom area accounted for in this visibility check? `${prefix}--modal-scroll-content` fades out the last part of the container, so an input can satisfy `elementRect.bottom <= modalRect.bottom` while still being visually obscured.

**Review by adamalston — COMMENTED** [review]
Could you add testing?

File: packages/react/src/components/Modal/Modal.tsx (hunk 1 above)
Comment (gDINESH13) [reply packages/react/src/components/Modal/Modal.tsx]: Oh ur right, I din thought of keyboard navigation accessibility.

The issue is distinguishing between:
1. **Click-initiated focus** (button clicked → receives focus → shouldn't scroll)
2. **Keyboard-initiated focus** (Tab key → button receives focus → SHOULD scroll)

Unfortunately, by the time the focus event fires, we can't reliably determine what caused it (click vs keyboard). The `FocusEvent` doesn't provide this information.

The intention initially for this change was, 

When the ComboBox clear button (X) was partially visible at the edge of the modal, clicking it would:

- Give the button focus
- Modal's focus handler would call scrollIntoView
- The scroll would happen during the click event
- The scroll interrupted the click, preventing the clear action
- User had to click again after scrolling to actually clear

The button check was meant to prevent this interruption by never scrolling buttons.
I will check if there are any better way to handle this. 

I thought of adding a *setTimeout* 
```
setTimeout(() => {
  currentActiveNode.scrollIntoView({ block: 'center' });
}, 10);
```
To  defer the scroll until after the click event completes. Do you think this would be a better approach?

File: packages/react/src/components/Modal/Modal.tsx (hunk 2 above)
Comment (gDINESH13) [reply packages/react/src/components/Modal/Modal.tsx:476]: You are right again 🫡.  The current visibility check doesn't account for the fade gradient. An element at the bottom edge can pass the check (`elementRect.bottom <= modalRect.bottom`) while being partially obscured by the fade.

I added this because even when combobox is completely visible like this, 
image

Clicking the button  triggers scroll, rather than showing dropdown options. 

Do you think adding some fixed margin or boundary 
```
const FADE_MARGIN = 48; // Adjust based on actual fade height

const isFullyVisible = 
  elementRect.top >= modalRect.top &&
  elementRect.bottom <= (modalRect.bottom - FADE_MARGIN);
```
This may not be the ideal way to handle this,
but this actually questions me back to the need of scroll into view, 

for scenarios like this, where button cancel in combobox is partially visible 
image

if I remove the button check you pointed above, it will scroll the element into view. 

Can you point me how much of an element should be visible in the modal for it to avoid scrolling?

File: packages/react/src/components/Modal/Modal.tsx (hunk 1 above)
Comment (adamalston) [reply packages/react/src/components/Modal/Modal.tsx]: Using `setTimeout` does not seem ideal. Nothing about this issue suggests that it is timing related.

File: packages/react/src/components/Modal/Modal.tsx (hunk 2 above)
Comment (adamalston) [reply packages/react/src/components/Modal/Modal.tsx:476]: My initial thought is that the entire element should be visible, as scrolling would not be necessary if it is fully in view. However, there may be other considerations (e.g., accessibility) that could influence the display. I didn't find any specific guidance in https://carbondesignsystem.com/components/modal/accessibility/ or https://carbondesignsystem.com/components/modal/style/.

Comment (gDINESH13) [reply packages/react/src/components/Modal/Modal.tsx:476]: Hi Adam, 
I tried to access the keyboard way of accessing the tabs below the visibility and the tabs which are not visible are actually scrolling into view even when (fade gradient ) is present.  
So I guess it does handle ```elementRect.bottom <= modalRect.bottom``` those tabs at bottom. 


https://github.com/user-attachments/assets/ea1b6ed0-c276-4ead-8c11-aa168819cee9

File: packages/react/src/components/Modal/Modal.tsx (hunk 1 above)
Comment (gDINESH13) [reply packages/react/src/components/Modal/Modal.tsx]: I have removed this entire button check of stopping the ```scrollintoview``` in modal. The combobox does clears its input with the current changes. When the element is completely into view. When combo-box is partially visibile like this, it scrolls the item into view before clearing first.
image

Do you think, this doesn't match the expected outcome?

**Review by tay1orjones — APPROVED** [review]
LGTM, I added a playwright e2e test to assert that scroll only happens when necessary. We can cover addl edge cases in subsequent PRs. We've had a couple folks reach out about this looking for a fix.

**Review by heloiselui — APPROVED** [review]

File: packages/react/src/components/Modal/Modal.tsx (hunk 2 above)
Comment (adamalston) [reply packages/react/src/components/Modal/Modal.tsx:476]: Okay, sounds good. I was going to suggest a change like to the following to fully account for the faded area. However, if what you have works, I think that's fine.

```diff
diff --git a/packages/react/src/components/Modal/Modal.tsx b/packages/react/src/components/Modal/Modal.tsx
index 9b978f5c75..1b5cf73047 100644
--- a/packages/react/src/components/Modal/Modal.tsx
+++ b/packages/react/src/components/Modal/Modal.tsx
@@ -463,13 +463,22 @@ const ModalDialog = React.forwardRef(function ModalDialog(
       return;
     }
 
+    const hasActiveScrollFade =
+      !modalContent.classList.contains(
+        `${prefix}--modal-scroll-content--no-fade`
+      ) && !modalContent.querySelector(`.${prefix}--autoalign`);
     const modalRect = modalContent.getBoundingClientRect();
+    /** Corresponds to the bottom fade distance in `_modal.scss`. */
+    const modalScrollFadeHeight = 80;
+    const visibleModalBottom = hasActiveScrollFade
+      ? modalRect.bottom - modalScrollFadeHeight
+      : modalRect.bottom;
     const elementRect = currentActiveNode.getBoundingClientRect();
 
     // Check if element is fully visible within the modal viewport
     const isFullyVisible =
       elementRect.top >= modalRect.top &&
-      elementRect.bottom <= modalRect.bottom;
+      elementRect.bottom <= visibleModalBottom;
 
     if (isFullyVisible) {
       return; // Don't scroll if already fully visible
```

**Review by adamalston — APPROVED** [review]

File: packages/react/src/components/Modal/Modal.tsx (hunk 2 above)
Comment (gDINESH13) [reply packages/react/src/components/Modal/Modal.tsx:476]: Yea thanks I kind of thought that too, if it din worked that was my next step.
